### PR TITLE
Prod 154 master

### PR DESCRIPTION
**Pull request by k-anderson**

### CODE REVIEW

**Comment by k-anderson** [comment]
@fenollp since tests currently run serially in the order of folder in core/ this will fail... kazoo_call needs kazoo_documents compiled for tests before it will succeed but they run in the opposite order.  Can you look into running make test-compile then the tests (or adding some way to do deps between apps) when running the kazoo test suite?

File: applications/callflow/doc/resources.md [hunk 1]
```diff
@@ -24,6 +24,14 @@ Key | Description | Type | Default | Required
 `to_did` | Statically set the DID to dial | `string` |   | `false`
 `use_local_resources` | Toggle whether to use the account's (or hunt_account_id's) resources vs the system resources | `boolean` | `true` | `false`
 
+## Special Dynamic Flags
+
+### Zone
+Adding the value "zone" to the dynamic_flags array will result in a flag set to the zone name that is executing the call.
+
+### Custom Channel Vars
+Adding the value "custom_channel_vars." appened with a normalized variable name will add the value (if present) to the flags array.  For example, "custom_channel_vars.owner_id".
```
Comment (jamesaimonetti) [inline applications/callflow/doc/resources.md]: s/appened/appended/

File: core/kazoo_endpoint/src/kz_attributes.erl [hunk 2]
```diff
@@ -562,6 +566,150 @@ owned_by_query(ViewOptions, Call, ViewKey) ->
...
+        'false' -> process_dynamic_flags(DynamicFlags, Flags, Call);
+        'true' ->
+            Fun = kz_term:to_atom(DynamicFlag),
+            Flag =  kapps_call:Fun(Call),
+            lager:debug("adding dynamic flag ~s", [Flag]),
+            process_dynamic_flags(DynamicFlags, [Flag|Flags], Call)
+    end.
+
+-spec is_flag_exported(ne_binary()) -> boolean().
+is_flag_exported(Flag) ->
```
Comment (fenollp) [inline core/kazoo_endpoint/src/kz_attributes.erl]: Let's match `?NE_BINARY` here to enforce our assumption

File: core/kazoo_endpoint/src/kz_endpoint.erl [hunk 3]
```diff
@@ -204,8 +204,9 @@ merge_attributes(Endpoint, Type) ->
            ,<<"mobile">>
            ,<<"presence_id">>
            ,<<"call_waiting">>
-           ,?ATTR_LOWER_KEY
            ,<<"formatters">>
+           ,<<"outbound_flags">>
+           ,?ATTR_LOWER_KEY
            ],
```
Comment (fenollp) [inline core/kazoo_endpoint/src/kz_endpoint.erl]: Let's put this in a /0 and test that it doesn't have duplicates.

File: core/kazoo_endpoint/src/kz_endpoint.erl [hunk 4]
```diff
@@ -249,6 +247,9 @@ merge_attribute(?ATTR_LOWER_KEY, _Account, Endpoint, Owner) ->
         'true' -> Endpoint;
         'false' -> kz_json:set_value(FullKey, OwnerAttr, Endpoint)
     end;
+merge_attribute(<<"call_restriction">>, Account, Endpoint, Owner) ->
+    Classifiers = kz_json:get_keys(knm_converters:available_classifiers()),
```
Comment (fenollp) [inline core/kazoo_endpoint/src/kz_endpoint.erl]: Shouldn't this look for the classifiers of `Account` specifically?

File: core/kazoo_endpoint/src/kz_endpoint.erl [hunk 5]
```diff
@@ -458,7 +469,7 @@ merge_call_restrictions([Classifier|Classifiers], Account, Endpoint, Owner) ->
 -spec get_user(ne_binary(), api_binary() | kz_json:object()) -> kz_json:object().
 get_user(_, 'undefined') -> kz_json:new();
 get_user(AccountDb, OwnerId) when is_binary(OwnerId) ->
```
Comment (fenollp) [inline core/kazoo_endpoint/src/kz_endpoint.erl]: Previous clause is caught by fetch

File: core/kazoo_endpoint/src/kz_endpoint.erl [hunk 6]
```diff
@@ -483,7 +494,7 @@ get_users(AccountDb, OwnerIds) ->
 get_users(_, [], Users) ->
     Users;
 get_users(AccountDb, [OwnerId|OwnerIds], Users) ->
-    case kz_datamgr:open_cache_doc(AccountDb, OwnerId) of
+    case kzd_user:fetch(AccountDb, OwnerId) of
```
Comment (fenollp) [inline core/kazoo_endpoint/src/kz_endpoint.erl]: I'd like to fill the cache beforehand with a bulk fetch. What do you think?

**Comment by fenollp** [comment]
I run your tests with `make compile-test && make test` and everything passes.
Travis does about the same thing and passes too.
That compile time dependency I will address in a future PR.

File: core/kazoo_endpoint/src/kz_attributes.erl [hunk 7]
```diff
@@ -562,6 +566,150 @@ owned_by_query(ViewOptions, Call, ViewKey) ->
...
+        [] -> Flags;
+        EndpointFlags -> EndpointFlags ++ Flags
+    end.
+
+-spec get_account_static_flags(ne_binary(), kapps_call:call(), ne_binaries()) ->
+                                      ne_binaries().
+get_account_static_flags(_, Call, Flags) ->
+    AccountId = kapps_call:account_id(Call),
+    case kz_account:fetch(AccountId) of
+        {'ok', AccountJObj} -> kz_device:outbound_static_flags(AccountJObj) ++ Flags;
```
Comment (fenollp) [inline core/kazoo_endpoint/src/kz_attributes.erl]: turns out it's ok

File: core/kazoo_endpoint/src/kz_attributes.erl [hunk 8]
```diff
@@ -562,6 +566,150 @@ owned_by_query(ViewOptions, Call, ViewKey) ->
...
+        DynamicFlags -> process_dynamic_flags(DynamicFlags, Flags, Call)
+    end.
+
+-spec get_account_dynamic_flags(ne_binary(), kapps_call:call(), ne_binaries()) ->
+                                       ne_binaries().
+get_account_dynamic_flags(_, Call, Flags) ->
+    AccountId = kapps_call:account_id(Call),
+    case kz_account:fetch(AccountId) of
+        {'ok', AccountJObj} ->
+            process_dynamic_flags(kz_device:outbound_dynamic_flags(AccountJObj), Flags, Call);
```
Comment (fenollp) [inline core/kazoo_endpoint/src/kz_attributes.erl]: turns out it's ok

File: core/kazoo_call/src/kapps_call.erl [hunk 9]
```diff
@@ -130,6 +130,10 @@
 
 -export([is_recording/1, set_is_recording/2]).
 
+-ifdef(TEST).
+-export([eq/2]).
```
Comment (fenollp) [inline core/kazoo_call/src/kapps_call.erl]: has to be here given the kapps_call record

File: core/kazoo_endpoint/src/kz_endpoint.erl (hunk 4 above)
Comment (fenollp) [inline core/kazoo_endpoint/src/kz_endpoint.erl]: nope

**Pull request merged**